File: test-apps/ui-test-app/android/imodeljs-test-app/app/src/main/java/com/bentley/imodeljs_test_app/MainActivity.java
package com.bentley.imodeljs_test_app;

import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.webkit.WebView;

import com.bentley.itwin.AuthCompletionAction;
import com.bentley.itwin.AuthSettings;
import com.bentley.itwin.AuthTokenCompletionAction;
import com.bentley.itwin.AuthorizationClient;
import com.bentley.itwin.IModelJsHost;
import com.bentley.itwin.MobileFrontend;

import java.time.Instant;

public class MainActivity extends AppCompatActivity {
    public class MyAuthClient extends AuthorizationClient {
        private final String tokenStr = "eyJhbGciOiJSUzI1NiIsImtpZCI6IkJlbnRsZXlJTVMiLCJwaS5hdG0iOiJhOG1lIn0.eyJzY29wZSI6WyJvcGVuaWQiLCJlbWFpbCIsInByb2ZpbGUiLCJvcmdhbml6YXRpb24iLCJpbW9kZWxodWIiLCJjb250ZXh0LXJlZ2lzdHJ5LXNlcnZpY2U6cmVhZC1vbmx5IiwicHJvZHVjdC1zZXR0aW5ncy1zZXJ2aWNlIiwicHJvamVjdHdpc2Utc2hhcmUiLCJ1cmxwcy10aGlyZC1wYXJ0eSIsImltb2RlbC1leHRlbnNpb24tc2VydmljZS1hcGkiLCJvZmZsaW5lX2FjY2VzcyJdLCJjbGllbnRfaWQiOiJpbW9kZWxqcy1lbGVjdHJvbi10ZXN0IiwiYXVkIjpbImh0dHBzOi8vaW1zLmJlbnRsZXkuY29tL2FzL3Rva2VuLm9hdXRoMiIsImh0dHBzOi8vaW1zb2lkYy5iZW50bGV5LmNvbS9hcy90b2tlbi5vYXV0aDIiLCJodHRwczovL2ltc29pZGMuYmVudGxleS5jb20vcmVzb3VyY2VzIiwiaW1vZGVsLWh1Yi1zZXJ2aWNlcy0yNDg1IiwiY29udGV4dC1yZWdpc3RyeS0yNzc3IiwicHJvZHVjdC1zZXR0aW5ncy1zZXJ2aWNlLTI3NTIiLCJwcm9qZWN0d2lzZS1zaGFyZS0yNTY3IiwidWxhcy1yZWFsdGltZS1sb2ctcG9zdGluZy0yNzMzIiwiaW1vZGVsLXBsdWctaW4tc2VydmljZS1hcGkiXSwic3ViIjoiYTBmZGJlODUtNzU0My00NWExLThmN2YtZGI3ZjQyZGI1YTlhIiwicm9sZSI6WyJNWV9TRUxFQ1RfQ0QiLCJQcm9qZWN0IE1hbmFnZXIiLCJTRUxFQ1RfRE9XTkxPQUQiLCJCRU5UTEVZX0VNUExPWUVFIl0sIm9yZyI6ImZhYjk3NzRiLWIzMzgtNGNjMi1hNmM5LTQ1OGJkZjdmOTY2YSIsInN1YmplY3QiOiJhMGZkYmU4NS03NTQzLTQ1YTEtOGY3Zi1kYjdmNDJkYjVhOWEiLCJpc3MiOiJodHRwczovL2ltc29pZGMuYmVudGxleS5jb20iLCJlbnRpdGxlbWVudCI6WyJCRE4iLCJTRUxFQ1RfMjAwNiIsIkJFTlRMRVlfTEVBUk4iLCJJTlRFUk5BTCJdLCJwcmVmZXJyZWRfdXNlcm5hbWUiOiJBZmZhbi5LaGFuQGJlbnRsZXkuY29tIiwiZ2l2ZW5fbmFtZSI6IkFmZmFuIiwic2lkIjoiNVNONWg1ZVhtUlI3bjhpWnZnZmpETnhDbXdBLlNVMVRMVUpsYm5Sc1pYa3RWVk0uMGZpVCIsIm5iZiI6MTYyNDM5NjMyMywidWx0aW1hdGVfc2l0ZSI6IjEwMDEzODkxMTciLCJ1c2FnZV9jb3VudHJ5X2lzbyI6IlVTIiwiYXV0aF90aW1lIjoxNjI0Mzk2NjIzLCJuYW1lIjoiQWZmYW4uS2hhbkBiZW50bGV5LmNvbSIsIm9yZ19uYW1lIjoiQmVudGxleSBTeXN0ZW1zIEluYyIsImZhbWlseV9uYW1lIjoiS2hhbiIsImVtYWlsIjoiQWZmYW4uS2hhbkBiZW50bGV5LmNvbSIsImV4cCI6MTYyNDYzOTU3NH0.bdIZI9rzOGV7VkpbVLomy7NkZ_h3aJ3yfHAZT-hZo6x_SLeSF_iap9cCdklBifq_1x7e9ooZGMNvIQIN3snR3KmB6HvXUaRyAppg_apjwcSrjLbuCU9LrMvBCktRRVngwj3WeXBDMBneX-QIYVEmwDV5IvrXtCxmD12hegYxxvvOy2Iw3avomvD_IMRvvRPP_OSkrzYoDPiUS2E1-9X54dqPa6v7ShdZrYaTTxAkyFD3j3rchDBwreQlu95IymU5JWx8Wy5qrhW_l2V-Po3f38RIKGq1pzLkwPDMp7XVych2eaXNlt2RwLM7NOouxLuREGDMM-p3czt_zg98XZmjsw";

        private String getTokenJson() {
            StringBuilder builder = new StringBuilder();
            final String tokenString = tokenStr;

            final String expiresAt = String.valueOf(Instant.now().toEpochMilli()+ 1000*60*60);
            final String startsAt= String.valueOf(Instant.now().toEpochMilli());
            builder.append("{");
            builder.append("\"tokenString\":").append("\"").append(tokenString).append("\",");
            builder.append("\"expiresAt\"  :").append(expiresAt).append(",");
            builder.append("\"startsAt\"   :").append(startsAt);
            builder.append("}");
            return builder.toString();
        }

        @Override
        public void initialize(AuthSettings authSettings, AuthCompletionAction authCompletionAction) {
            try {
                authCompletionAction.resolve();
            } catch (Exception e) {
                Log.e("%s", e.getMessage());
            }
        }

        @Override
        public void signIn(AuthCompletionAction authCompletionAction) {
            try {
                notifyUserStateChangedWithToken(getTokenJson());
                authCompletionAction.resolve();
            } catch (Exception e) {
                Log.e("%s", e.getMessage());
            }
        }

        @Override
        public void signOut(AuthCompletionAction authCompletionAction) {
            try {
                notifyUserStateChangedWithToken("");
                authCompletionAction.resolve();
            } catch (Exception e) {
                Log.e("%s", e.getMessage());
            }
        }

        @Override
        public void getAccessToken(AuthTokenCompletionAction authTokenCompletionAction) {
            try {
                authTokenCompletionAction.resolve(getTokenJson());
            } catch (Exception e) {
                Log.e("%s", e.getMessage());
            }
        }

        @Override
        public boolean isAuthorized() {
            return true;
        }
    };

    IModelJsHost m_host;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        WebView.setWebContentsDebuggingEnabled(true);
        boolean alwaysExtractAssets = true; // for debugging, otherwise the host will only extract when app version changes
        m_host = new IModelJsHost(this, alwaysExtractAssets, new MyAuthClient(), true );
        m_host.startup();
        String files = getFilesDir().getPath();
        MobileFrontend frontend = new MobileFrontend(m_host, "");
        frontend.getSettings().setAllowFileAccessFromFileURLs(true);;
        frontend.getSettings().setAllowUniversalAccessFromFileURLs(true);;

        m_host.setFrontend(frontend);
        frontend.loadEntryPoint();
        setContentView(frontend);
    }

    @Override
    protected void onPause() {
        super.onPause();
        m_host.onPause();
    }

    @Override
    protected void onResume() {
        super.onResume();
        m_host.onResume();
    }
}

File: test-apps/display-test-app/android/imodeljs-test-app/app/src/main/java/com/bentley/imodeljs_test_app/MainActivity.kt
/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
package com.bentley.imodeljs_test_app

import android.app.Activity
import android.content.ContentResolver
import android.content.Context
import android.content.Intent
import android.content.res.AssetManager
import android.net.Uri
import android.os.Bundle
import android.provider.OpenableColumns
import android.webkit.JavascriptInterface
import android.webkit.WebResourceRequest
import android.webkit.WebResourceResponse
import android.webkit.WebView
import androidx.activity.result.contract.ActivityResultContract
import androidx.appcompat.app.AppCompatActivity
import androidx.webkit.WebViewAssetLoader
import androidx.webkit.WebViewClientCompat
import com.bentley.itwin.IModelJsHost
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.net.URLConnection
import kotlin.system.exitProcess
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import org.json.JSONObject

/**
 * Determines the display name for the uri.
 * @param uri The input content Uri.
 * @return The display name or null if it wasn't found.
 */
fun ContentResolver.getFileDisplayName(uri: Uri): String? {
    // Query the content resolver only if we have a content Uri
    return uri.takeIf { uri.scheme == ContentResolver.SCHEME_CONTENT }?.let {
        query(uri, arrayOf(OpenableColumns.DISPLAY_NAME), null, null, null, null)?.use { cursor ->
            cursor.takeIf { cursor.moveToFirst() }?.getString(0)
        }
    }
}

/**
 * Copies the input uri to the destination directory.
 * @param uri The input content Uri to copy.
 * @param destDir The destination directory.
 * @return The full path of the copied file, null if it failed.
 */
fun Context.copyToExternalFiles(uri: Uri, destDir: String): String? {
    return contentResolver.getFileDisplayName(uri)?.let { displayName ->
        getExternalFilesDir(null)?.let { filesDir ->
            contentResolver.openInputStream(uri)?.use { input ->
                val outDir = File(filesDir.path, destDir)
                outDir.mkdirs()
                val outputFile = File(outDir, displayName)
                outputFile.outputStream().use {
                    input.copyTo(it)
                }
                outputFile.path
            }
        }
    }
}

fun JSONObject.optStringNotEmpty(name: String): String? {
    return optString(name).takeIf { it.isNotEmpty() }
}

typealias PickUriContractType = ActivityResultContract<Nothing?, Uri?>

class PickUriContract(private val destDir: String) : PickUriContractType() {
    private lateinit var context: Context

    override fun createIntent(context: Context, input: Nothing?): Intent {
        this.context = context
        return Intent()
            .setAction(Intent.ACTION_OPEN_DOCUMENT)
            .setType("*/*")
            .addCategory(Intent.CATEGORY_OPENABLE)
    }

    override fun parseResult(resultCode: Int, intent: Intent?): Uri? {
        val uri = intent?.takeIf { resultCode == Activity.RESULT_OK }?.data
        if (uri != null) {
            context.copyToExternalFiles(uri, destDir)?.let { result ->
                return Uri.parse(result)
            }
        }
        return uri
    }
}

// Very similar to WebViewAssetLoader.AssetsPathHandler except it doesn't log an error when a resource isn't found, also doesn't handle SvgZ streams.
class AssetsPathHandler(context: Context) : WebViewAssetLoader.PathHandler {
    private val mContext = context

    private fun removeLeadingSlash(path: String): String {
        return if (path.startsWith('/')) path.substring(1) else path
    }

    private fun openAsset(path: String): InputStream {
        return mContext.assets.open("www/${removeLeadingSlash(path)}", AssetManager.ACCESS_STREAMING)
    }

    private fun guessMimeType(path: String): String {
        return URLConnection.guessContentTypeFromName(path) ?: "text/plain"
    }

    override fun handle(path: String): WebResourceResponse? {
        return try {
            WebResourceResponse(guessMimeType(path), null, openAsset(path))
        } catch (e: IOException) {
            WebResourceResponse(null, null, null)
        }
    }
}

@Suppress("SpellCheckingInspection")
class MainActivity : AppCompatActivity() {
    private lateinit var host: IModelJsHost
    private var promiseName: String = ""
    private lateinit var env: JSONObject
    private var exitAfterModelOpened = false

    companion object {
        const val BIM_CACHE_DIR = "bim_cache"
    }

    private fun log(message: String) {
        // Print out a message with a unique prefix so that it can be uniquely found in the TORRENT
        // of text that comes out of adb logcat.
        println("com.bentley.display_test_app: $message")
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WebView.setWebContentsDebuggingEnabled(true)
        loadEnvJson()
        val authClient = DtaServiceAuthorizationClient.create(env)
        val alwaysExtractAssets = true // for debugging, otherwise the host will only extract when app version changes
        host = IModelJsHost(this, alwaysExtractAssets, authClient, true).apply {
            setBackendPath("www/mobile")
            setHomePath("www/home")
            startup()
        }

        val webView = WebView(this)
        // using a WebViewAssetLoader so that the localization json files load properly
        // the version of i18next-http-backend we're using tries to use the fetch API with file URL's (apparently fixed in version 2.0.1)
        val assetLoader = WebViewAssetLoader.Builder().addPathHandler("/", AssetsPathHandler(this)).build()

        webView.webViewClient = object : WebViewClientCompat() {
            override fun shouldInterceptRequest(view: WebView?, request: WebResourceRequest): WebResourceResponse? {
                return assetLoader.shouldInterceptRequest(request.url)
            }
        }

        val startPickDocument = registerForActivityResult(PickUriContract(BIM_CACHE_DIR)) { uri ->
            if (uri != null && promiseName.isNotEmpty()) {
                val js = "if (window.$promiseName) window.$promiseName(\"$uri\"); else console.log('Error: window.$promiseName is not defined!');"
                MainScope().launch {
                    webView.evaluateJavascript(js, null)
                }
            }
        }

        // The linter is broken, and claims that the interface functions in this class do not have
        // the @JavascriptInterface annotation, despite the fact that they do.
        @Suppress("JavascriptInterface")
        webView.addJavascriptInterface(
            object {
                @JavascriptInterface
                @Suppress("unused")
                fun openModel(promiseName: String) {
                    this@MainActivity.promiseName = promiseName
                    startPickDocument.launch(null)
                }

                @JavascriptInterface
                @Suppress("unused")
                fun modelOpened(modelName: String) {
                    log("iModel opened: $modelName")
                }

                @JavascriptInterface
                @Suppress("unused", "UNUSED_PARAMETER")
                fun firstRenderFinished(_dummy: String) {
                    log("First render finished.")
                    if (exitAfterModelOpened) {
                        exitProcess(0)
                    }
                }
            }, "DTA_Android")

        host.webView = webView
        setContentView(webView)

        var args = "&standalone=true"
        env.optStringNotEmpty("IMJS_STANDALONE_FILENAME")?.let { fileName ->
            // ensure fileName already exists in the external files
            getExternalFilesDir(BIM_CACHE_DIR)?.let { filesDir ->
                val fullPath = File(filesDir, fileName)
                if (fullPath.exists()) {
                    args += "&iModelName=${Uri.encode(fullPath.toString())}"
                } else {
                    log("requested imodel ($fullPath) not found!")
                }
            }
        }

        if (authClient != null) {
            val remoteIds = listOfNotNull(
                env.optStringNotEmpty("IMJS_ITWIN_ID"),
                env.optStringNotEmpty("IMJS_IMODEL_ID")
            )
            if (remoteIds.size == 2) {
                args += "&iTwinId=${Uri.encode(remoteIds[0])}&iModelId=${Uri.encode(remoteIds[1])}"
            }
        }

        if (isYesEnv("IMJS_IGNORE_CACHE"))
            args += "&ignoreCache=true"

        if (isYesEnv("IMJS_EXIT_AFTER_MODEL_OPENED"))
            exitAfterModelOpened = true

        host.loadEntryPoint(env.optStringNotEmpty("IMJS_DEBUG_URL") ?: "https://${WebViewAssetLoader.DEFAULT_DOMAIN}/index.html", args)
    }

    private fun isYesEnv(name: String): Boolean {
        return env.optString(name) == "YES"
    }

    private fun loadEnvJson() {
        env = try {
            JSONObject(assets.open("www/mobile/env.json").bufferedReader().use { it.readText() })
        } catch (ex: Exception) {
            JSONObject()
        }
    }

    override fun onResume() {
        host.onResume()
        super.onResume()
    }

    override fun onPause() {
        host.onPause()
        super.onPause()
    }
}